File: src/drms/core/vector_store.py
"""
Vector Store implementation using ChromaDB for document embeddings and search.
"""

import asyncio
import logging
from typing import List, Dict, Any, Optional
import chromadb
from chromadb.config import Settings
import openai
from sentence_transformers import SentenceTransformer

logger = logging.getLogger(__name__)

class VectorStore:
    def __init__(self, 
                 db_path: str = "./chroma_db",
                 embedding_model: str = "all-MiniLM-L6-v2",
                 use_openai_embeddings: bool = False,
                 openai_api_key: Optional[str] = None):
        """
        Initialize vector store with ChromaDB backend.
        
        Args:
            db_path: Path to ChromaDB persistent storage
            embedding_model: Sentence transformer model name
            use_openai_embeddings: Whether to use OpenAI embeddings
            openai_api_key: OpenAI API key if using OpenAI embeddings
        """
        self.db_path = db_path
        self.use_openai_embeddings = use_openai_embeddings
        
        # Initialize ChromaDB
        self.client = chromadb.PersistentClient(
            path=db_path,
            settings=Settings(anonymized_telemetry=False)
        )
        
        # Initialize embedding model
        if use_openai_embeddings and openai_api_key:
            openai.api_key = openai_api_key
            self.embedding_model = None
        else:
            # Force CPU usage to avoid MPS issues
            import torch
            device = "cpu"  # Force CPU to avoid MPS crashes
            self.embedding_model = SentenceTransformer(embedding_model, device=device)
        
        # Create collections for different types of documentation
        self.collections = {
            "docs": self._get_or_create_collection("documentation"),
            "api": self._get_or_create_collection("api_reference"),
            "examples": self._get_or_create_collection("code_examples"),
            "tutorials": self._get_or_create_collection("tutorials")
        }
        
        logger.info(f"VectorStore initialized with {len(self.collections)} collections")
    
    def _get_or_create_collection(self, name: str):
        """Get or create a ChromaDB collection."""
        try:
            return self.client.get_collection(name)
        except:
            return self.client.create_collection(
                name=name,
                metadata={"hnsw:space": "cosine"}
            )
    
    async def _get_embeddings(self, texts: List[str]) -> List[List[float]]:
        """Generate embeddings for texts."""
        if self.use_openai_embeddings:
            response = await openai.Embedding.acreate(
                model="text-embedding-ada-002",
                input=texts
            )
            return [item["embedding"] for item in response["data"]]
        else:
            # Use sentence transformers in a thread to avoid blocking
            loop = asyncio.get_event_loop()
            embeddings = await loop.run_in_executor(
                None, self.embedding_model.encode, texts
            )
            return embeddings.tolist()
    
    async def add_documents(self, 
                          documents: List[Dict[str, Any]], 
                          collection_type: str = "docs") -> bool:
        """
        Add documents to the vector store.
        
        Args:
            documents: List of document dicts with 'content', 'metadata', 'id'
            collection_type: Type of collection ('docs', 'api', 'examples', 'tutorials')
        
        Returns:
            bool: Success status
        """
        try:
            if collection_type not in self.collections:
                logger.error(f"Unknown collection type: {collection_type}")
                return False
            
            collection = self.collections[collection_type]
            
            # Prepare data
            texts = [doc["content"] for doc in documents]
            ids = [doc["id"] for doc in documents]
            metadatas = [doc.get("metadata", {}) for doc in documents]
            
            # Generate embeddings
            embeddings = await self._get_embeddings(texts)
            
            # Add to collection
            collection.add(
                embeddings=embeddings,
                documents=texts,
                metadatas=metadatas,
                ids=ids
            )
            
            logger.info(f"Added {len(documents)} documents to {collection_type} collection")
            return True
            
        except Exception as e:
            logger.error(f"Error adding documents: {e}")
            return False
    
    async def search_documents(self, 
                             query: str, 
                             collection_type: str = "docs",
                             n_results: int = 5,
                             filter_metadata: Optional[Dict] = None) -> List[Dict[str, Any]]:
        """
        Search for similar documents.
        
        Args:
            query: Search query
            collection_type: Collection to search
            n_results: Number of results to return
            filter_metadata: Metadata filters
        
        Returns:
            List of search results with scores
        """
        try:
            if collection_type not in self.collections:
                logger.error(f"Unknown collection type: {collection_type}")
                return []
            
            collection = self.collections[collection_type]
            
            # Generate query embedding
            query_embedding = await self._get_embeddings([query])
            
            # Search
            results = collection.query(
                query_embeddings=query_embedding,
                n_results=n_results,
                where=filter_metadata,
                include=["documents", "metadatas", "distances"]
            )
            
            # Format results
            formatted_results = []
            for i in range(len(results["documents"][0])):
                formatted_results.append({
                    "content": results["documents"][0][i],
                    "metadata": results["metadatas"][0][i],
                    "similarity": 1 - results["distances"][0][i],  # Convert distance to similarity
                    "id": results["ids"][0][i] if "ids" in results else f"result_{i}"
                })
            
            logger.info(f"Found {len(formatted_results)} results for query: {query[:50]}...")
            return formatted_results
            
        except Exception as e:
            logger.error(f"Error searching documents: {e}")
            return []
    
    async def search_multi_collection(self, 
                                    query: str, 
                                    n_results: int = 3) -> Dict[str, List[Dict[str, Any]]]:
        """Search across all collections and return categorized results."""
        results = {}
        
        for collection_name in self.collections.keys():
            results[collection_name] = await self.search_documents(
                query=query,
                collection_type=collection_name,
                n_results=n_results
            )
        
        return results
    
    async def get_collection_stats(self) -> Dict[str, Dict[str, Any]]:
        """Get statistics for all collections."""
        stats = {}
        
        for name, collection in self.collections.items():
            try:
                count = collection.count()
                stats[name] = {
                    "document_count": count,
                    "collection_name": name
                }
            except Exception as e:
                stats[name] = {"error": str(e)}
        
        return stats
    
    async def delete_documents(self, 
                             doc_ids: List[str], 
                             collection_type: str = "docs") -> bool:
        """Delete documents by IDs."""
        try:
            if collection_type not in self.collections:
                return False
            
            collection = self.collections[collection_type]
            collection.delete(ids=doc_ids)
            
            logger.info(f"Deleted {len(doc_ids)} documents from {collection_type}")
            return True
            
        except Exception as e:
            logger.error(f"Error deleting documents: {e}")
            return False
    
    async def update_document(self, 
                            doc_id: str, 
                            content: str, 
                            metadata: Dict[str, Any],
                            collection_type: str = "docs") -> bool:
        """Update a single document."""
        try:
            # Delete old version
            await self.delete_documents([doc_id], collection_type)
            
            # Add new version
            return await self.add_documents([{
                "id": doc_id,
                "content": content,
                "metadata": metadata
            }], collection_type)
            
        except Exception as e:
            logger.error(f"Error updating document {doc_id}: {e}")
            return False

File: src-node/core/vector-store.ts
/**
 * Vector Store implementation using ChromaDB for document embeddings and search.
 */

import { ChromaClient, OpenAIEmbeddingFunction } from 'chromadb';
import { OpenAI } from 'openai';
import { pipeline, env } from '@xenova/transformers';
import { getSettings } from '../config/settings.js';

// Configure transformers to use local models
env.allowLocalModels = true;
env.allowRemoteModels = true;

export interface Document {
  id: string;
  content: string;
  metadata: Record<string, any>;
}

export interface SearchResult {
  content: string;
  metadata: Record<string, any>;
  similarity: number;
  id: string;
}

export interface CollectionStats {
  documentCount: number;
  collectionName: string;
}

export class VectorStore {
  private client!: ChromaClient;
  private collections: Record<string, any> = {};
  private embeddingFunction: any;
  private transformerPipeline: any;
  private settings = getSettings();

  constructor() {
    this.initializeClient();
    this.initializeEmbeddingFunction();
  }

  private initializeClient(): void {
    this.client = new ChromaClient({
      path: this.settings.vectorDbPath
    });
  }

  private async initializeEmbeddingFunction(): Promise<void> {
    if (this.settings.useOpenaiEmbeddings && this.settings.openaiApiKey) {
      // Use OpenAI embeddings
      this.embeddingFunction = new OpenAIEmbeddingFunction({
        openai_api_key: this.settings.openaiApiKey,
        openai_model: 'text-embedding-ada-002'
      });
    } else {
      // Use default embedding function (no need for DefaultEmbeddingFunction import)
      this.embeddingFunction = undefined;
      // Initialize the transformer pipeline for local embeddings
      try {
        this.transformerPipeline = await pipeline('feature-extraction', this.settings.embeddingModel);
      } catch (error) {
        console.warn('Failed to load local embedding model, falling back to default:', error);
      }
    }
  }

  public async initialize(): Promise<void> {
    try {
      // Create collections for different types of documentation
      const collectionNames = ['documentation', 'api_reference', 'code_examples', 'tutorials'];
      
      for (const name of collectionNames) {
        try {
          this.collections[name] = await this.client.getCollection({
            name,
            embeddingFunction: this.embeddingFunction
          });
        } catch {
          // Collection doesn't exist, create it
          this.collections[name] = await this.client.createCollection({
            name,
            embeddingFunction: this.embeddingFunction,
            metadata: { 'hnsw:space': 'cosine' }
          });
        }
      }

      console.log(`VectorStore initialized with ${Object.keys(this.collections).length} collections`);
    } catch (error) {
      console.error('Failed to initialize VectorStore:', error);
      throw error;
    }
  }

  private async generateEmbeddings(texts: string[]): Promise<number[][]> {
    if (this.settings.useOpenaiEmbeddings && this.settings.openaiApiKey) {
      // OpenAI embeddings are handled by the embedding function
      return [];
    } else if (this.transformerPipeline) {
      // Use local transformer pipeline
      const embeddings = await Promise.all(
        texts.map(async (text) => {
          const result = await this.transformerPipeline(text, { pooling: 'mean', normalize: true });
          return Array.from(result.data as Float32Array);
        })
      );
      return embeddings;
    } else {
      // Fallback to default embedding function
      return [];
    }
  }

  public async addDocuments(
    documents: Document[],
    collectionType: string = 'documentation'
  ): Promise<boolean> {
    try {
      const collection = this.collections[collectionType];
      if (!collection) {
        console.error(`Unknown collection type: ${collectionType}`);
        return false;
      }

      const texts = documents.map(doc => doc.content);
      const ids = documents.map(doc => doc.id);
      const metadatas = documents.map(doc => doc.metadata);

      // Generate embeddings if using local model
      let embeddings: number[][] | undefined;
      if (!this.settings.useOpenaiEmbeddings) {
        embeddings = await this.generateEmbeddings(texts);
      }

      await collection.add({
        ids,
        documents: texts,
        metadatas,
        embeddings: embeddings?.length ? embeddings : undefined
      });

      console.log(`Added ${documents.length} documents to ${collectionType} collection`);
      return true;
    } catch (error) {
      console.error('Error adding documents:', error);
      return false;
    }
  }

  public async searchDocuments(
    query: string,
    collectionType: string = 'documentation',
    nResults: number = 5,
    filterMetadata?: Record<string, any>
  ): Promise<SearchResult[]> {
    try {
      const collection = this.collections[collectionType];
      if (!collection) {
        console.error(`Unknown collection type: ${collectionType}`);
        return [];
      }

      // Generate query embedding if using local model
      let queryEmbedding: number[] | undefined;
      if (!this.settings.useOpenaiEmbeddings && this.transformerPipeline) {
        const result = await this.transformerPipeline(query, { pooling: 'mean', normalize: true });
        queryEmbedding = Array.from(result.data);
      }

      const results = await collection.query({
        queryTexts: queryEmbedding ? undefined : [query],
        queryEmbeddings: queryEmbedding ? [queryEmbedding] : undefined,
        nResults,
        where: filterMetadata,
        include: ['documents', 'metadatas', 'distances']
      });

      // Format results
      const formattedResults: SearchResult[] = [];
      if (results.documents && results.documents[0]) {
        for (let i = 0; i < results.documents[0].length; i++) {
          formattedResults.push({
            content: results.documents[0][i] || '',
            metadata: results.metadatas?.[0]?.[i] || {},
            similarity: 1 - (results.distances?.[0]?.[i] || 0), // Convert distance to similarity
            id: results.ids?.[0]?.[i] || `result_${i}`
          });
        }
      }

      console.log(`Found ${formattedResults.length} results for query: ${query.substring(0, 50)}...`);
      return formattedResults;
    } catch (error) {
      console.error('Error searching documents:', error);
      return [];
    }
  }

  public async searchMultiCollection(
    query: string,
    nResults: number = 3
  ): Promise<Record<string, SearchResult[]>> {
    const results: Record<string, SearchResult[]> = {};

    for (const collectionName of Object.keys(this.collections)) {
      results[collectionName] = await this.searchDocuments(
        query,
        collectionName,
        nResults
      );
    }

    return results;
  }

  public async getCollectionStats(): Promise<Record<string, CollectionStats | { error: string }>> {
    const stats: Record<string, CollectionStats | { error: string }> = {};

    for (const [name, collection] of Object.entries(this.collections)) {
      try {
        const count = await collection.count();
        stats[name] = {
          documentCount: count,
          collectionName: name
        };
      } catch (error) {
        stats[name] = { error: String(error) };
      }
    }

    return stats;
  }

  public async deleteDocuments(
    docIds: string[],
    collectionType: string = 'documentation'
  ): Promise<boolean> {
    try {
      const collection = this.collections[collectionType];
      if (!collection) {
        return false;
      }

      await collection.delete({
        ids: docIds
      });

      console.log(`Deleted ${docIds.length} documents from ${collectionType}`);
      return true;
    } catch (error) {
      console.error('Error deleting documents:', error);
      return false;
    }
  }

  public async updateDocument(
    docId: string,
    content: string,
    metadata: Record<string, any>,
    collectionType: string = 'documentation'
  ): Promise<boolean> {
    try {
      // Delete old version
      await this.deleteDocuments([docId], collectionType);

      // Add new version
      return await this.addDocuments([{
        id: docId,
        content,
        metadata
      }], collectionType);
    } catch (error) {
      console.error(`Error updating document ${docId}:`, error);
      return false;
    }
  }

  public getCollectionNames(): string[] {
    return Object.keys(this.collections);
  }

  public async close(): Promise<void> {
    // ChromaDB doesn't require explicit closing in the current API
    console.log('VectorStore closed');
  }
}